### PR TITLE
rbd/cache: Replicated Write Log core codes - aio_discard

### PR DESCRIPTION
**Pull request by yison**
Adds Replicated Write Log, a persistent, write back cache for Ceph RBD.
Implements: http://tracker.ceph.com/projects/ceph/wiki/Rbd_-_ordered_crash-consistent_write-back_caching_extension

Trello: https://trello.com/c/QnsQaGTn

[Obsoletes https://github.com//pull/29087]

The PR 29087 is split into smaller PRs to make review easier.
The first PR: #31279
The second PR: #31963
The 3rd PR: #33502
The 4th PR: #34430
The 5th PR: #34699
The 6th PR: #34706 

This is the 7th PR (aio_discard) which defines the following parts:
librbd: add DiscardLogEntry
librbd: add DiscardLogOperation
librbd: add DiscardRequest
librbd: add aio_discard
librbd: add aio_discard test case

Signed-off-by: Scott Peterson scott.d.peterson@intel.com
Signed-off-by: Lisa Li xiaoyan.li@intel.com
Signed-off-by: Yuan Lu yuan.y.lu@intel.com
Signed-off-by: Mahati Chamarthy mahati.chamarthy@intel.com

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by yison** [comment]
@dillaman Hi Jason. This PR is about aio_discard. This is the 7th PR of Replicated Write Log core codes. Please review it when you're convenient. Thanks very much.

File: src/librbd/cache/rwl/LogEntry.cc [hunk 1]
```diff
@@ -163,6 +163,35 @@ std::ostream &operator<<(std::ostream &os,
   return entry.format(os);
 }
 
+void DiscardLogEntry::writeback(librbd::cache::ImageWritebackInterface &image_writeback,
+                                Context *ctx) {
+  image_writeback.aio_discard(ram_entry.image_offset_bytes, ram_entry.write_bytes,
+                              1, ctx);
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogEntry.cc:169]: Nit: you pass the discard granularity to `C_DiscardRequest` but then don't use it. It should be stored in the log entry.

File: src/librbd/cache/rwl/LogEntry.h [hunk 2]
```diff
@@ -86,7 +86,7 @@ class GenericWriteLogEntry : public GenericLogEntry {
   ~GenericWriteLogEntry() override {};
   GenericWriteLogEntry(const GenericWriteLogEntry&) = delete;
   GenericWriteLogEntry &operator=(const GenericWriteLogEntry&) = delete;
-  inline unsigned int write_bytes() {
+  virtual inline unsigned int write_bytes() {
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogEntry.h:89]: Nit:
```suggestion
  virtual unsigned int write_bytes() const {
```

File: src/librbd/cache/rwl/LogEntry.h [hunk 3]
```diff
@@ -159,6 +159,35 @@ class WriteLogEntry : public GenericWriteLogEntry {
...
+    : GenericWriteLogEntry(sync_point_entry, image_offset_bytes, write_bytes) {
+    ram_entry.discard = 1;
+  };
+  DiscardLogEntry(const uint64_t image_offset_bytes, const uint64_t write_bytes)
+    : GenericWriteLogEntry(nullptr, image_offset_bytes, write_bytes) {
+    ram_entry.discard = 1;
+  };
+  DiscardLogEntry(const DiscardLogEntry&) = delete;
+  DiscardLogEntry &operator=(const DiscardLogEntry&) = delete;
+  inline unsigned int write_bytes() override {
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogEntry.h:175]: Nit: 
```suggestion
  unsigned int write_bytes() const override {
```

File: src/librbd/cache/rwl/LogEntry.h [hunk 4]
```diff
@@ -159,6 +159,35 @@ class WriteLogEntry : public GenericWriteLogEntry {
...
+    /* The valid bytes in this ops data buffer. */
+    return 0;
+  };
+  inline unsigned int bytes_dirty() override {
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogEntry.h:179]: Nit: (fix the const-ness of the virtual method)
```suggestion
  unsigned int bytes_dirty() const override {
```

File: src/librbd/cache/rwl/LogOperation.h [hunk 5]
```diff
@@ -50,6 +50,9 @@ class GenericLogOperation {
   virtual bool reserved_allocated() {
     return false;
   }
+  virtual bool is_writing_op() {
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogOperation.h:53]: Nit:
```suggestion
  virtual bool is_writing_op() const {
```

File: src/librbd/cache/rwl/LogOperation.h [hunk 6]
```diff
@@ -166,6 +172,35 @@ class WriteLogOperationSet {
...
+                      const utime_t dispatch_time,
+                      PerfCounters *perfcounter,
+                      CephContext *cct);
+  ~DiscardLogOperation() override;
+  DiscardLogOperation(const DiscardLogOperation&) = delete;
+  DiscardLogOperation &operator=(const DiscardLogOperation&) = delete;
+  const std::shared_ptr<GenericLogEntry> get_log_entry() override {
+    return log_entry;
+  }
+  bool reserved_allocated() override {
```
Comment (dillaman) [inline src/librbd/cache/rwl/LogOperation.h:194]: Nit: (fix const-ness of virtual method)
```suggestion
  bool reserved_allocated() const override {
```

File: src/librbd/cache/rwl/LogEntry.cc (hunk 1 above)
Comment (yison) [reply src/librbd/cache/rwl/LogEntry.cc]: Done

File: src/librbd/cache/rwl/LogEntry.h (hunk 2 above)
Comment (yison) [reply src/librbd/cache/rwl/LogEntry.h]: Done

File: src/librbd/cache/rwl/LogEntry.h (hunk 3 above)
Comment (yison) [reply src/librbd/cache/rwl/LogEntry.h]: Done

File: src/librbd/cache/rwl/LogEntry.h (hunk 4 above)
Comment (yison) [reply src/librbd/cache/rwl/LogEntry.h]: Done

File: src/librbd/cache/rwl/LogOperation.h (hunk 5 above)
Comment (yison) [reply src/librbd/cache/rwl/LogOperation.h]: Done

File: src/librbd/cache/rwl/LogOperation.h (hunk 6 above)
Comment (yison) [reply src/librbd/cache/rwl/LogOperation.h]: Done

**Comment by yison** [comment]
@dillaman Hi Jason, all comments are solved.

**Comment by yison** [comment]
@dillaman Hi Jason, I rebased latest ceph master to fix conflicts.

**Pull request merged**